elastic/integrations
# [Kubernetes] Add support for TSDB - all metrics datastreams except the state ones

## What does this PR do?
Sets the dimension fields for each metric datastream, except the `state_*` ones (for now), so they can migrate to TSDB.

## Why is this important?
The fields set as they are now, are not sufficient to migrate to TSDB. We want this migration so we can create "a space efficient, easy to use, and faster solution for storing metrics data generated by Elastic’s solutions and 3rd party data providers." (more details on the goal of this change can be found [here](https://github.com/elastic/elasticsearch-adrs/blob/master/analytics/tsdb/tsdb-design.md#project-statement)).

## Details on this PR

Important considerations:
1. Each document from a TSDB index has a `_tsid` that distinguishes it from the other documents. This ID is generated based on the **timestamp** and all **dimension** fields of the document.
2. Each index has a default capacity for 16 dimension fields, **but** 8 of these are reserved for `ecs` fields. It is possible to change this default (check [here](https://github.com/elastic/integrations/blob/main/docs/developer_tsdb_migration_guidelines.md#-steps-for-migrating-an-existing-package)).
3. Dimension values cannot go over 1024b.
4. Not all field types classify for dimension (check which ones [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds.html#time-series-dimension)).
5. Each dimension field is part of the [`index.routing_path`](https://www.elastic.co/guide/en/elasticsearch/reference/current/tsds-index-settings.html#index-routing-path). Every document is only "accepted" in a TSDB index if **there is at least one dimension field** present.
6. If a metric is split by labels, then the **labels should be set as dimension**. Otherwise, only one of two (or more) documents with the same metric for a different label will be stored.

Main changes:
1. Every datastream has the keyword fields `service.address` and `orchestrator.cluster.url` set to dimensions.
2. In combination to that:
   1. Pod: `pod.uid`, as it is unique across the whole cluster.
   2. Node: no more are necessary.
   3. Container: `pod.uid` (same reason as for Pod).
   4. System: `kubernetes.system.container`, since it is a metric label and documents may be being split on it.
   5. Volume: `kubernetes.volume.name` (metric label), `kubernetes.pod.name` and `kubernetes.namespace`, as the pod name is unique per namespace.
   6. Controller manager, API server, proxy and scheduler: all labels are set to dimensions.

Important warning: TSDB is not enabled by default (check _Screenshots_ below on how). Visualizations with counter metric fields are unavailable as of the time of the creation of this PR - problem described [here](https://github.com/elastic/kibana/issues/152467).

## Example

Th `_tsid` for a TSDB Kubernetes system datastream is obtained through the combination of the timestamp and the dimension fields `kubernetes.system.container`,  `service.address` and `orchestrator.cluster.url`. In an example consisting of an Elastic Agent in a two node clusters we could have the following documents:
![image](https://user-images.githubusercontent.com/113898685/223454179-c7eef708-7c2b-4b3d-8cf1-2a6e77e19a7d.png)
In this document, every field has the same value except for the `kubernetes.system.container`. If this field was not set to dimension, then the `_tsid` generated would be the same for the two documents and a conflict would happen.

For a more hands on example, check the section _TSDB very simple example_ in https://github.com/elastic/integrations/issues/4618.

## Checklist



## How to test this PR locally

1. Clone this repository.
2. Run `elastic-package build` inside Kubernetes package.
3. Deploy the Elastic stack.
4. Check the number of documents in each index before enabling the TSDB and after enabling it (check _Screenshots_ to see how to enable it).

## Related issues

- Relates https://github.com/elastic/integrations/issues/4618


## Screenshots

General view of the documents of all these datastreams on **Discover**:


![Screenshot from 2023-03-07 15-03-33](https://user-images.githubusercontent.com/113898685/223452416-f58830e0-15be-4856-8519-0d5f0a06f433.png)

From this screenshot, there was no change in the number of documents before and after enabling TSDB.

To enable TSDB, all that is necessary to do is to go to agent Policy, Kubernetes integration and enable TSDB through the toggle available when clicking "Advanced options" under a datastream:
![image](https://user-images.githubusercontent.com/113898685/223452978-35a9b517-7b12-4e3e-89f5-dac416fbea49.png)
